lib/utils/platform/iot_threads_afr.c

```diff
@@ -0,0 +1,350 @@
+    }
+    
+    pMutex->recursive = recursive; /* remember the type of mutex */
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Destroy( IotMutex_t * const pMutex )
+{
+    _iot_mutex_internal_t * internalMutex = ( _iot_mutex_internal_t * ) pMutex;
+
+    vSemaphoreDelete( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+}
+
+/*-----------------------------------------------------------*/
+
+bool prIotMutexTimedLock(IotMutex_t * const pMutex, TickType_t timeout)
+{
+    _iot_mutex_internal_t * internalMutex = ( _iot_mutex_internal_t * ) pMutex;
+    BaseType_t  lockResult;
+
+    IotLogDebug( "Locking mutex %p.", internalMutex );
+
+    /* Call the correct FreeRTOS mutex take function based on mutex type. */
+    if( internalMutex->recursive == true )
+    {
+        lockResult = xSemaphoreTakeRecursive( ( SemaphoreHandle_t ) & internalMutex->xMutex, timeout );
+    }
+    else
+    {
+        lockResult = xSemaphoreTake( ( SemaphoreHandle_t ) & internalMutex->xMutex, timeout);
+    }
+  
+    return ( lockResult == pdTRUE );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Lock( IotMutex_t * const pMutex )
+{
+    prIotMutexTimedLock( pMutex, portMAX_DELAY );
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotMutex_TryLock( IotMutex_t * const pMutex )
+{
+    return prIotMutexTimedLock( pMutex, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Unlock( IotMutex_t * const pMutex )
+{
+    _iot_mutex_internal_t * internalMutex = ( _iot_mutex_internal_t * ) pMutex;
+
+    IotLogDebug( "Unlocking mutex %p.", internalMutex );
+    
+    /* Call the correct FreeRTOS mutex unlock function based on mutex type. */
+    if( internalMutex->recursive == true )
+    {
+        ( void ) xSemaphoreGiveRecursive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+    else
+    {
+         ( void ) xSemaphoreGive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_Create( IotSemaphore_t * const pNewSemaphore,
+                             uint32_t initialValue,
+                             uint32_t maxValue )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pNewSemaphore;
+    IotLogDebug( "Creating new semaphore %p.", pNewSemaphore );
+
+    ( void ) xSemaphoreCreateCountingStatic( maxValue, initialValue, &internalSemaphore->xSemaphore );
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+uint32_t IotSemaphore_GetCount( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+    UBaseType_t count = 0;
+
+    count = ( UBaseType_t ) uxSemaphoreGetCount( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore );
```

Comment:
No need to case to `UBaseType_t` here (return value is of matching type).

Reply:
Agree